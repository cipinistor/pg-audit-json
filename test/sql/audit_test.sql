CREATE EXTENSION "pg-audit-json";
SET client_min_messages TO WARNING;

CREATE TABLE foo (
  id BIGSERIAL PRIMARY KEY,
  a INTEGER,
  b INTEGER,
  c INTEGER
);

SELECT audit.audit_table('foo', true, '{id}'::text[]);

--
-- Basic INSERT/UPDATE/DELETE
--

--- insert
INSERT INTO foo (a, b, c) VALUES (1, 2, 3) RETURNING id AS record_id \gset
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--- update
UPDATE foo SET b = 99 WHERE id = :record_id;
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--- delete
DELETE FROM foo WHERE id = :record_id;
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--
-- With table modifications
--

--- insert
INSERT INTO foo (a, b, c) VALUES (4, 5, 6) RETURNING id AS record_id \gset
SELECT rpk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--- add column to table and update record
ALTER TABLE foo ADD COLUMN d INTEGER;
UPDATE foo SET d = 7 WHERE id = :record_id;
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--- remove column from table and update record
ALTER TABLE foo DROP COLUMN d;
UPDATE foo SET c = 99 WHERE id = :record_id;
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

--- delete
DELETE FROM foo WHERE id = :record_id;
SELECT pk, old, new FROM audit.log ORDER BY id DESC LIMIT 1;

DROP TABLE foo;

DROP EXTENSION "pg-audit-json";
