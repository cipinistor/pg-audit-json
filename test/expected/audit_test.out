CREATE EXTENSION "pg-audit-json";
SET client_min_messages TO WARNING;
CREATE TABLE foo (
  id BIGSERIAL PRIMARY KEY,
  a INTEGER,
  b INTEGER,
  c INTEGER
);
SELECT audit.audit_table('foo', true, '{id}'::text[]);
 audit_table 
-------------
 
(1 row)

--
-- Basic INSERT/UPDATE/DELETE
--
--- insert
INSERT INTO foo (a, b, c) VALUES (1, 2, 3) RETURNING id AS record_id \gset
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  | old_values | new_values 
-----------+------------+------------
 {"id": 1} |            | 
(1 row)

--- update
UPDATE foo SET b = 99 WHERE id = :record_id;
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  | old_values | new_values 
-----------+------------+------------
 {"id": 1} | {"b": 2}   | {"b": 99}
(1 row)

--- delete
DELETE FROM foo WHERE id = :record_id;
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  |        old_values         | new_values 
-----------+---------------------------+------------
 {"id": 1} | {"a": 1, "b": 99, "c": 3} | 
(1 row)

--
-- With table modifications
--
--- insert
INSERT INTO foo (a, b, c) VALUES (4, 5, 6) RETURNING id AS record_id \gset
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  | old_values | new_values 
-----------+------------+------------
 {"id": 2} |            | 
(1 row)

--- add column to table and update record
ALTER TABLE foo ADD COLUMN d INTEGER;
UPDATE foo SET d = 7 WHERE id = :record_id;
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  | old_values  | new_values 
-----------+-------------+------------
 {"id": 2} | {"d": null} | {"d": 7}
(1 row)

--- remove column from table and update record
ALTER TABLE foo DROP COLUMN d;
UPDATE foo SET c = 99 WHERE id = :record_id;
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  | old_values | new_values 
-----------+------------+------------
 {"id": 2} | {"c": 6}   | {"c": 99}
(1 row)

--- delete
DELETE FROM foo WHERE id = :record_id;
SELECT table_pk, old_values, new_values FROM audit.log ORDER BY id DESC LIMIT 1;
 table_pk  |        old_values         | new_values 
-----------+---------------------------+------------
 {"id": 2} | {"a": 4, "b": 5, "c": 99} | 
(1 row)

DROP TABLE foo;
DROP EXTENSION "pg-audit-json";
